### ISSUE TITLE
Possible connection leak when used outside unit of work

### ISSUE THREAD

**Issue by GuiSim** [supercargo/guice-persist-jooq#9]
When you use the `Provider<DSLContext>` from guice-persist-jooq, you call `.get()` to get the transaction-aware `DSLContext` and start using it.

When calling get, the provider will `begin()` the Unit of work (UoW).

```

  public DSLContext get() {
    if (!isWorking()) {
      begin();
    }

    DSLContext factory = threadFactory.get();
    Preconditions.checkState(null != factory, "Requested Factory outside work unit. "
        + "Try calling UnitOfWork.begin() first, or use a PersistFilter if you "
        + "are inside a servlet environment.");

    return factory;
  }
``` 

Sadly, there's no way that this UoW will get closed unless it is by "luck".

In my opinion, components responsible for starting a UoW should be responsible for closing them. I think this leads to clean design.

The scenario that we hit from time to time is that we forget one code path with `@Transactional`. This means that there are no `MethodInterceptor` on this code path and thus the UoW is not started when we `get()` the `DSLContext`. This starts the UoW but *nothing will call `end()`* since we're not in a `@Transactional` method.

The result is connections that are never `closed()` and thus never released back to our connection pool.

The clean solution to me is to change `get()` to do this:

```
  public DSLContext get() {
    DSLContext factory = threadFactory.get();
    Preconditions.checkState(null != factory, "Requested Factory outside work unit. "
        + "Try calling UnitOfWork.begin() first, or use a PersistFilter if you "
        + "are inside a servlet environment.");

    return factory;
  }
``` 

This ensures that a UoW was started before we get the `DSLContext`. This will fail fast and ensures errors are caught early in the development cycle.

I'd like your thoughts on this. It is a somewhat major change but I think it will only lead to better code from users of this library.

It is also a potentially breaking change. I can imagine that some users have their connection pool to release checkedout connections after a certain time. This would cause the lib to not leak (but only through a sub-optimal mechanism). My proposed change would break this currently supported scenario.

Another option is to add some sort of `strict` parameter that would not `begin()` the UoW when calling `get()`. I'd rather not have that but if you feel like this is the safest approach, I can do this.

Let me know what you prefer and I'll get a PR.

**Comment by supercargo** [comment]
Hi @GuiSim, thanks for taking the time to write this up!

I agree with the general direction you are proposing, but do have a couple of concerns around the breaking nature.  I think there may be a bigger issue than users who are doing some side-cleanup in their connection pool.  It seems like there would be a code change for any library consumer using `Provider<DSLContext>` but not the annotations.  Transaction control is in the hand of the user here, and that code currently assumes calls to `Provider<DSLContext>#get()` start a transaction if needed.  Another potential issue is that `begin()` isn't in the `Provider` interface.

For those reasons I do think a high level configuration flag would be warranted...just need to determine if default would be strict (breaking) or current (non-breaking).

**Comment by GuiSim** [comment]
Thanks for the fast reply! I really appreciate it.

>  It seems like there would be a code change for any library consume using Provider<DSLContext> but not the annotations. 

Do you have an example of that in mind? Like you said, `begin()` and `end()` are not on the `Provider` interface so how would they start and more importantly end these UoW? 

One way is through `PersistFilter`, but that fits with my suggestion of enforcing "UoW's `begin` caller should be the same as the `end()` caller" so no breaking changes here.

**Comment by supercargo** [comment]
Yes, good point.  Sorry it has been a while since I've delved into this code.  I guess I'm still searching for ways in which the loss of auto-starting would be a problem/surprise to users of the library.  If there are none then simply removing those lines would probably be best.  When using Guice scopes for similar things (e.g. tenant scoped dependancies in a multi-tenant application) I've had a lot of success with the fail-fast approach you're proposing.

I'd say go ahead and put this into a PR.  An option to restore the original behavior (non-default) would still be nice just because I'd prefer to be a little bit conservative with breaking changes.  We could even mark that configuration option as deprecated from the outset and eliminate it in future versions.

**Comment by GuiSim** [comment]
Thanks @supercargo 

I created a PR. I decided not to add an option to restore the original behavior but if you insist I will add it.

Let me know what you think!

**Comment by supercargo** [comment]
This is fixed in 1.0.0 which is published to central.  There may be a delay of several hours before it is available on the mirrors.
